SimCalorimetry/HcalSimAlgos/src/HcalSiPM.cc

```diff
@@ -12,53 +12,88 @@
+    : theCellCount(nCells),
+      theSiPM(nCells, -999.f),
+      theCrossTalk(0.),
+      theTempDep(0.),
+      theLastHitTime(-1.),
+      nonlin(nullptr) {
   setTau(tau);
   assert(theCellCount > 0);
-  resetSiPM();
 }
 
 HcalSiPM::~HcalSiPM() {
   if (nonlin)
     delete nonlin;
 }
 
-//================================================================================
-//implementation of Borel-Tanner distribution
-double HcalSiPM::Borel(unsigned int n, double lambda, unsigned int k) {
-  if (n < k)
-    return 0;
-  double dn = double(n);
-  double dk = double(k);
-  double dnk = dn - dk;
-  double ldn = lambda * dn;
-  double logb = -ldn + dnk * log(ldn) - TMath::LnGamma(dnk + 1);
-  double b = 0;
-  if (logb >= -20) {  // protect against underflow
-    b = (dk / dn);
-    if ((n - k) < 100)
-      b *= (exp(-ldn) * pow(ldn, dnk)) / TMath::Factorial(n - k);
-    else
-      b *= exp(logb);
+namespace {
+
+  /*
+  //================================================================================
+  //original implementation of Borel-Tanner distribution
+  // here for reference
+  constexpr double originalBorel(unsigned int n, double lambda, unsigned int k) {
+    if (n < k)
+      return 0;
+    double dn = double(n);
+    double dk = double(k);
+    double dnk = dn - dk;
+    double ldn = lambda * dn;
+    double logb = -ldn + dnk * log(ldn) - TMath::LnGamma(dnk + 1);
+    double b = 0;
+    if (logb >= -20) {  // protect against underflow
+      b = (dk / dn);
+      if ((n - k) < 100)
+        b *= (exp(-ldn) * pow(ldn, dnk)) / TMath::Factorial(n - k);
+      else
+        b *= exp(logb);
+    }
+    return b;
+  }
+  */
+
+  using FLOAT = double;
+  //================================================================================
+  //modified implementation of Borel-Tanner distribution
+  constexpr double Borel(unsigned int i, FLOAT lambda, unsigned int k, double iFact) {
+    auto n = k + i;
+    FLOAT dn = FLOAT(n);
+    FLOAT dk = FLOAT(k);
+    FLOAT dnk = FLOAT(i);
+
+    FLOAT ldn = lambda * dn;
+    FLOAT b0 = (dk / dn);
+    FLOAT b = 0;
+    if (i < 100) {
+      b = b0 * (std::exp(-ldn) * std::pow(ldn, dnk)) / iFact;
+    } else {
+      FLOAT logb = -ldn + dnk * std::log(ldn) - std::log(iFact);
+      // protect against underflow
+      b = (logb >= -20.) ? b0 * std::exp(logb) : 0;
+    }
+    return b;
   }
-  return b;
-}
+
+}  // namespace
 
 const HcalSiPM::cdfpair& HcalSiPM::BorelCDF(unsigned int k) {
   // EPSILON determines the min and max # of xtalk cells that can be
   // simulated.
-  static const double EPSILON = 1e-6;
-  typename cdfmap::const_iterator it;
-  it = borelcdfs.find(k);
+  constexpr double EPSILON = 1e-6;
+  auto it = borelcdfs.find(k);
   if (it == borelcdfs.end()) {
     vector<double> cdf;
+    cdf.reserve(64);
 
     // Find the first n=k+i value for which cdf[i] > EPSILON
     unsigned int i;
-    double b = 0., sumb = 0.;
+    double sumb = 0.;
+    double iFact = 1.;
     for (i = 0;; i++) {
```

Comment:
```suggestion
    for (i = 0; i<171; i++) {
```
Maybe picky, but I guess the same issue _could_ also happen in this previous loop.